zipline/src/jsMain/kotlin/app/cash/zipline/ziplineFunctionJs.kt

```diff
@@ -0,0 +1,139 @@
+ */
+@ExperimentalSerializationApi
+fun <T : ZiplineService> ZiplineFunction<T>.asDynamicFunction(): (service: T, args: List<*>) -> Any? {
+  require(this is ReturningZiplineFunction<T>) {
+    "asDynamicFunction() cannot be called on $this"
+  }
+
+  val resultSerializer = ResultOrCallbackSerializer(FastDynamicSerializer)
+  val argsListSerializer = ArgsListSerializer(
+    serializers = List(argsListSerializer.serializers.size) { FastDynamicSerializer },
+  )
+
+  return { service, args ->
+    val callHandler = (service as OutboundService).callHandler
+    callHandler.callInternal(
+      service = service,
+      function = this@asDynamicFunction,
+      argsListSerializer = argsListSerializer,
+      resultSerializer = resultSerializer,
+      args = args.toTypedArray(),
+    )
+  }
+}
+
+/**
+ * Returns a suspending function that accepts and returns plain JavaScript objects, instead of
+ * regular Kotlin values. (The function's receiver will still receive regular Kotlin values as
+ * parameters, and return a regular Kotlin value as a result.)
+ *
+ * Use this in highly performance-sensitive code to reduce the amount of work required to call
+ * bridged functions on Zipline services.
+ */
+@ExperimentalSerializationApi
+fun <T : ZiplineService> ZiplineFunction<T>.asDynamicSuspendingFunction(): suspend (service: T, args: List<*>) -> Any? {
+  require(this is SuspendingZiplineFunction<T>) {
+    "asDynamicSuspendingFunction() cannot be called on $this"
+  }
+
+  val resultOrCallbackSerializer = ResultOrCallbackSerializer(FastDynamicSerializer)
+  val argsListSerializer = ArgsListSerializer(
+    serializers = List(argsListSerializer.serializers.size) { FastDynamicSerializer },
+  )
+  val dynamicSuspendCallbackSerializer = ziplineServiceSerializer<SuspendCallback<Any?>>(
+    kClass = SuspendCallback::class,
+    typeArgumentsSerializers = listOf(FastDynamicSerializer),
+  )
+
+  return { service, args ->
+    val callHandler = (service as OutboundService).callHandler
+    callHandler.callSuspendingInternal(
+      service = service,
+      function = this@asDynamicSuspendingFunction,
+      argsListSerializer = argsListSerializer,
+      resultOrCallbackSerializer = resultOrCallbackSerializer,
+      suspendCallbackSerializer = dynamicSuspendCallbackSerializer,
+      args = args.toTypedArray(),
+    )
+  }
+}
+
+/**
+ * ‘Encode’ JavaScript objects as JSON by assigning them without any intermediate copies.
+ *
+ * **This is not a general-purpose serializer** and can only be used with
+ * `Json.encodeToStringFast()`.
+ *
+ * In particular, this assumes that the target [Encoder] is building a JavaScript object tree and
+ * not a string or stream. It assumes [AbstractEncoder.encodeValue] will accept a JavaScript object
+ * directly. It also assumes the output JSON tree will be used once and discarded, and skips the
+ * deep copy of input values.
+ */
+@ExperimentalSerializationApi
+internal object FastDynamicSerializer : KSerializer<Any?> {
```

Comment:
Kinda wanna promote this to kotlinx.serialization. Here’s an early attempt at that.
https://github.com/swankjesse/kotlinx.serialization/pull/1